### ISSUE TITLE
[Quest] Better user feedback incase the app does not have internet connection  

### ISSUE THREAD

**Issue by f-odhiambo** [opensrp/fhircore#2283]
**Describe the bug**

**To Reproduce**

**Expected behavior**

**Screenshots**

**Smartphone (please complete the following information):**
 - Device: [e.g. Samsung Galaxy A2 emulator]
 - Android version: [e.g. Settings -> About phone -> Android version]
 - Build number: [e.g. Settings -> About phone -> Build number]
 - Version [e.g. 24]

**Additional context**

![image](https://user-images.githubusercontent.com/4540684/234272713-cb473d05-826e-498c-85b7-a47173ade0e5.png)

**Comment by f-odhiambo** [comment]
Should we make the message configurable? 
CC @ndegwamartin @ellykits

**Comment by ndegwamartin** [comment]
> Should we make the message configurable? CC @ndegwamartin @ellykits

Not sure it needs to be configurable but it should at least be translatable

**Comment by ndegwamartin** [comment]
@HenryRae @rowo do have any thoughts on an appropriate user feedback messages for the above scenario?

**Comment by ndegwamartin** [comment]
@f-odhiambo we are now logging the server side exception so that it is available to the developer in debug mode. 

For _screenshot 1_ that message remains for the UI. It is best practice to keep such server side error messages away from the user and only show user friendly/simple ones.

For the second error we can have the error message read as either of these:

1. **Login Error: Please enter the same username had used initially**
2. **Login Error: Attempted to login in with a different user**
3. **Login Error: This user account is not recognized on this device**
4. **Login Error: Sorry we could't find an account with that username on this device**
5. **Login Error: Username not recognized on this device**
5. **Login Error: That account does not exist on this device. Enter a different account**

cc @HenryRae

**Comment by f-odhiambo** [comment]
> @f-odhiambo we are now logging the server side exception so that it is available to the developer in debug mode.
> 
> For _screenshot 1_ that message remains for the UI. It is best practice to keep such server side error messages away from the user and only show user friendly/simple ones.
> 
> For the second error we can have the error message read as either of these:
> 
> 1. **Login Error: Please enter the same username had used initially**
> 2. **Login Error: Attempted to login in with a different user**
> 3. **Login Error: This user account is not recognized on this device**
> 4. **Login Error: Sorry we could't find an account with that username on this device**
> 5. **Login Error: Username not recognized on this device**
> 6. **Login Error: That account does not exist on this device. Enter a different account**
> 
> cc @HenryRae

Let's go with option 1

**Comment by f-odhiambo** [comment]
> This is an error that shows if the user tries to log in with a different user name after having already logged in with one user name, is that correct? Why can't they do that?

Yes, that is the case. So based on the current implementation , if I logged in using `Demo` I cannot username username to log back in incase the app session expires

**Comment by pld** [comment]
Ah ok, this is b/c of offline login I guess? In case, we should make the username not editable while they're offline

**Comment by ndegwamartin** [comment]
I think it would happen both online and offline, whenever you try to log in with an unrecognised username.

**Comment by pld** [comment]
Isn't unrecognized username different, @f-odhiambo you were describing this as a different username, not an unrecognized username. If this about unrecognized username, then an error makes sense, like "Your username or password is incorrect"

**Comment by ndegwamartin** [comment]
> Isn't unrecognized username different, @f-odhiambo you were describing this as a different username, not an unrecognized username. If this about unrecognized username, then an error makes sense, like "Your username or password is incorrect"

It is unrecognized in the context of the device. It could be a user existing in the identity provider but the device wouldn't know that if it was offline.

**Comment by pld** [comment]
K, I think we need to talk this through with @rowo, we do know that the user is offline so we could have the UX take that into account

**Comment by pld** [comment]
I've chatted w/Roger, I think we have a plan, based on the answer to this question:
- If User A logs in while online
- Then User B logs in while online
- Then the device goes offline
- User B logs out
- Can User A log back in?

If the answer to the above is NO. Then we should disable changing the username while the device is offline. But on tap, show the user a message, like: `Please connect to the internet to change your username.`

**Comment by ndegwamartin** [comment]
Currently, the app does not support multi tenancy whether per provider or per team _(or other sync strategy)_, any new user logging in would need to perform a data purge and fresh install

**Comment by ndegwamartin** [comment]
Currently, the app does not support multi tenancy whether per provider or per team (or other sync strategy), any new user logging in would need to perform a data purge and fresh install. This goes for whether the user is online or offline.

**Comment by ndegwamartin** [comment]
> Cool, so then the fix here is as described in the above message, I've updated the description and title.

Do you mean this message? `Please connect to the internet to change your username.` 

Also note the username cannot be changed when online as well unless one does a fresh install (data wipe)

**Comment by pld** [comment]
Oh right, that's a very good point. So perhaps we should disallow changing the username regardless of whether they are online or offline. Does that make sense to you?

**Comment by ndegwamartin** [comment]
> 

Yeah that sounds good

**Comment by ndegwamartin** [comment]
Yeah that sounds good - at least until we reinstate multi-tenancy which we probably will at some point

**Comment by ndegwamartin** [comment]
> K, I update the description to address this. This is kinda a big issue, do we have an issue to allow the user to log in (when online) with a different username?
> 
> I think it's kinda reasonable to do that issue first.

I think it might already be somewhere on the roadmap cc @dubdabasoduba @f-odhiambo

**Comment by dubdabasoduba** [comment]
> > K, I update the description to address this. This is kinda a big issue, do we have an issue to allow the user to log in (when online) with a different username?
> > I think it's kinda reasonable to do that issue first.
> 
> I think it might already be somewhere on the roadmap cc @dubdabasoduba @f-odhiambo

Yes, this is added with the title **Multi-user login on the same device**

**Comment by pld** [comment]
Can you share a link to that? I can't find it in this repo

**Comment by ndegwamartin** [comment]
It is also worth noting that @LZRS had picked up one half of this in the overall OpenSRP roadmap but won't probably move as fast because of priority on client work. That is the ticket here https://github.com/onaio/opensrp-tasks/issues/85

**Comment by pld** [comment]
Yep cool, I'm am (personally, but this is for discussion and not decided) OK leaving this issue open until we have a first round of multi-tenancy merged.

@LZRS if you wanna take on a first round of multi-tenancy for FHIR core, you can start by writing up a new ticket in this repo, some scope it will have to cover
1. only allows user to change the username when online
2. if a user has unsynced data, do we disallow a different user from logging in until the user with unsynced data has logged in and synced? -- I'm OK not doing this, if we have some very clear warning about data loss, in fact, we probably should have a very clear warning about data loss whenever there is unsynced data and the current user is logged out (even without multitenancy), because if the user follows the current errr recommended (!?) approach and deletes the app, all that data's going to be gone anyhow)

**Comment by HenryRae** [comment]
I think this copy can further be improve. My apologies for late reply 
> Please connect to the internet to change your username.

1. `You attempted to login to this device with a different username. Please connect to the internet to login with different username.`
2. `To login to this device with different username, please connect to the internet`
3. `To login with different username, please connect to the internet` 

I would suggest option 3. 
cc: @ndegwamartin @rowo

**Comment by pld** [comment]
Based on the above discussion that functionality doesn't exist right now, so if we want to work on this before we implement multi-user, we would lock the username and not allow it to be changed after logging in once, regardless of connectivity.
